File: pluralsight/applyingfunctionalprincipalsincsharp/refactoringawayfromexceptions/ExceptionsAndReadability.Web/Infrastructure/Result.cs
namespace ExceptionsAndReadability.Web.Infrastucture
{
    public class Result
    {
        public bool IsSuccess { get; }

        public bool IsFailure => !IsSuccess;

        public string Error { get; }

        private Result(bool isSuccess, string error)
        {
            IsSuccess = isSuccess;
            Error = error;
        }

        public static Result Ok()
        {
            return new Result(true, string.Empty);
        }

        public static Result Fail(string error)
        {
            return new Result(false, error);
        }
    }
}

File: alura/course_android_room_001/app/src/main/java/br/com/flaviogf/schedule/infrastructure/Result.java
package br.com.flaviogf.schedule.infrastructure;

public class Result<T> {
    private final T value;
    private final Boolean isSuccess;
    private final String message;

    private Result(T value, Boolean isSuccess, String message) {
        this.value = value;
        this.isSuccess = isSuccess;
        this.message = message;
    }

    public T getValue() {
        return value;
    }

    public Boolean isSuccess() {
        return isSuccess;
    }

    public Boolean isFailure() {
        return !isSuccess;
    }

    public String getMessage() {
        return message;
    }

    public static <T> Result<T> ok(T value) {
        return new Result<>(value, true, null);
    }

    public static <T> Result<T> fail(String message) {
        return new Result<>(null, false, message);
    }
}
